src/socket.rs

```diff
@@ -62,27 +66,98 @@ impl Socket for UdpSocket {
 
         Ok(Self {
             socket,
-            complete_address
+            complete_address,
         })
     }
 
     fn apply_timeout(&self, timeout_settings: Option<TimeoutSettings>) -> GDResult<()> {
         let settings = timeout_settings.unwrap_or_default();
-        self.socket.set_read_timeout(settings.get_read()).unwrap();   //unwrapping because TimeoutSettings::new
+        self.socket.set_read_timeout(settings.get_read()).unwrap(); //unwrapping because TimeoutSettings::new
         self.socket.set_write_timeout(settings.get_write()).unwrap(); //checks if these are 0 and throws an error
 
         Ok(())
     }
 
     fn send(&mut self, data: &[u8]) -> GDResult<()> {
-        self.socket.send_to(data, &self.complete_address).map_err(|_| PacketSend)?;
+        self.socket
+            .send_to(data, &self.complete_address)
+            .map_err(|_| PacketSend)?;
         Ok(())
     }
 
     fn receive(&mut self, size: Option<usize>) -> GDResult<Vec<u8>> {
         let mut buf: Vec<u8> = vec![0; size.unwrap_or(DEFAULT_PACKET_SIZE)];
-        let (number_of_bytes_received, _) = self.socket.recv_from(&mut buf).map_err(|_| PacketReceive)?;
+        let (number_of_bytes_received, _) =
+            self.socket.recv_from(&mut buf).map_err(|_| PacketReceive)?;
 
         Ok(buf[..number_of_bytes_received].to_vec())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::thread;
+
+    use super::*;
+
+    #[test]
+    fn test_tcp_socket_send_and_receive() {
+        // Spawn a thread to run the server
+        let server_thread = thread::spawn(|| {
+            let listener = net::TcpListener::bind("127.0.0.1:8080").unwrap();
+            let (mut stream, _) = listener.accept().unwrap();
+            let mut buf = [0; 1024];
+            stream.read(&mut buf).unwrap();
+            stream.write(&buf).unwrap();
+        });
+
+        // Create a TCP socket and send a message to the server
+        let mut socket = TcpSocket::new("127.0.0.1", 8080).unwrap();

```

Comment:
It seems that it hangs on both tcp and udp sockets while being created, I have added `thread::sleep(Duration::from_millis(200));` before these (because I guess they try to connect and the listener from the newly made threads aren't fully initialized yet) and I can't get them to hang anymore.

Reply:
Or better yet, by initializing the listener sockets outside the newly made threads, then move them in (via `move ||`), it works as expected now and without delays.

Reply:
sounds good ill put a commit together